test/gtest/uct/test_uct_perf.cc

```diff
@@ -15,8 +15,64 @@ extern "C" {
 #define MB                        pow(1024, -2)
 #define UCT_PERF_TEST_MULTIPLIER  5
 #define UCT_ARM_PERF_TEST_MULTIPLIER  15
+#define UCT_CUDA_PERF_TEST_MULTIPLIER 10
 
 class test_uct_perf : public uct_test, public test_perf {
+public:
+    void test_execute(unsigned flags = 0) {
+        if (has_transport("ugni_udt")) {
+            UCS_TEST_SKIP;
+        }
+
+        /* For SandyBridge CPUs, don't check performance of far-socket devices */
+        std::vector<int> cpus = get_affinity();
+        bool check_perf       = true;
+        size_t max_iter       = std::numeric_limits<size_t>::max();
+
+        if (ucs_arch_get_cpu_model() == UCS_CPU_MODEL_INTEL_SANDYBRIDGE) {
+            for (std::vector<int>::iterator iter = cpus.begin();
+                 iter != cpus.end(); ++iter) {
+                if (!ucs_cpu_is_set(*iter, &GetParam()->local_cpus)) {
+                    UCS_TEST_MESSAGE << "Not enforcing performance on "
+                                        "SandyBridge far socket";
+                    check_perf = false;
+                    break;
+                }
+            }
+        }
+
+        if (has_transport("tcp")) {
+            check_perf = false; /* TODO calibrate expected performance based on transport */
+            max_iter   = 1000lu;
+        }
+
+        /* Run all tests */
+        for (const test_spec *test_iter = tests; test_iter->title != NULL;
+             ++test_iter) {
+            test_spec test = *test_iter;
+
+            test.send_mem_type = UCS_MEMORY_TYPE_HOST;
+            test.recv_mem_type = UCS_MEMORY_TYPE_HOST;
+
+            if (has_transport("cuda_copy") || has_transport("gdr_copy")) {
+                test.max *= UCT_CUDA_PERF_TEST_MULTIPLIER;
+                test.min /= UCT_CUDA_PERF_TEST_MULTIPLIER;
+
+                test.recv_mem_type = UCS_MEMORY_TYPE_CUDA;
```

Comment:
maybe the memory type could be specified by another test class for cuda-loopback transports?
to avoid has_transport check

something like:
```
class test_uct_loopback_cuda : public test_uct_perf {
};

UCS_TEST_P(test_uct_loopback_cuda , envelope)
{
    test_execute(UCS_MEMORY_TYPE_CUDA, UCT_CUDA_PERF_TEST_MULTIPLIER, 
                 UCX_PERF_TEST_FLAG_LOOPBACK);
}

UCT_INSTANTIATE_TEST_CASE_TLS(test_uct_loopback, UCT_TEST_CUDA_MEM_TYPE_TLS)
```

(add new macro `UCT_INSTANTIATE_TEST_CASE_TLS(_suite, _tls)`)

Reply:
We already have a lot usage of "has_transport" - even in the same function... Create a separate class instead of one "if" statement ? - don't think it's a right thing...

Reply:
BTW, why changing only test.recv_mem_type and not also send_mem_type?

Reply:
Because gdr_copy requires host mem for send buff

Reply:
and cuda_copy? sounds we want to test cuda_copy with 3 combinations: host->cuda, cuda->host, cuda->cuda

Reply:
host->cuda, cuda->host - already tested as put/get
cuda->cuda ? The instance of the class (the test) cannot be host->cuda & cuda->cuda at the same time